File: src/com/kingmang/lazurite/parser/AST/Expressions/UnaryExpression.java
package com.kingmang.lazurite.parser.AST.Expressions;

import com.kingmang.lazurite.exceptions.LZRException;
import com.kingmang.lazurite.parser.AST.Statements.Statement;
import com.kingmang.lazurite.parser.AST.Accessible;
import com.kingmang.lazurite.patterns.visitor.ResultVisitor;
import com.kingmang.lazurite.patterns.visitor.Visitor;
import com.kingmang.lazurite.runtime.Types.LzrNumber;
import com.kingmang.lazurite.runtime.Types.LzrString;
import com.kingmang.lazurite.core.Types;
import com.kingmang.lazurite.runtime.LzrValue;


public final class UnaryExpression implements Expression, Statement {

    public enum Operator {
        INCREMENT_PREFIX("++"),
        DECREMENT_PREFIX("--"),
        INCREMENT_POSTFIX("++"),
        DECREMENT_POSTFIX("--"),
        NEGATE("-"),
        // Boolean
        NOT("!"),
        // Bitwise
        COMPLEMENT("~");
        
        private final String name;

        private Operator(String name) {
            this.name = name;
        }

        @Override
        public String toString() {
            return name;
        }
    }
    
    public final Expression expr1;
    public final Operator operation;

    public UnaryExpression(Operator operation, Expression expr1) {
        this.operation = operation;
        this.expr1 = expr1;
    }
    
    @Override
    public void execute() {
        eval();
    }
    
    @Override
    public LzrValue eval() {
        final LzrValue value = expr1.eval();
        switch (operation) {
            case INCREMENT_PREFIX: {
                if (expr1 instanceof com.kingmang.lazurite.parser.AST.Accessible) {
                    return ((com.kingmang.lazurite.parser.AST.Accessible) expr1).set(increment(value));
                }
                return increment(value);
            }
            case DECREMENT_PREFIX: {
                if (expr1 instanceof com.kingmang.lazurite.parser.AST.Accessible) {
                    return ((com.kingmang.lazurite.parser.AST.Accessible) expr1).set(decrement(value));
                }
                return decrement(value);
            }
            case INCREMENT_POSTFIX: {
                if (expr1 instanceof com.kingmang.lazurite.parser.AST.Accessible) {
                    ((com.kingmang.lazurite.parser.AST.Accessible) expr1).set(increment(value));
                    return value;
                }
                return increment(value);
            }
            case DECREMENT_POSTFIX: {
                if (expr1 instanceof com.kingmang.lazurite.parser.AST.Accessible) {
                    ((Accessible) expr1).set(decrement(value));
                    return value;
                }
                return decrement(value);
            }
            case NEGATE: return negate(value);
            case COMPLEMENT: return complement(value);
            case NOT: return not(value);
            default:
                throw new LZRException("OperationIsNotSupportedException ","Operation " + operation + " is not supported");
        }
    }
    
    private LzrValue increment(LzrValue value) {
        if (value.type() == Types.NUMBER) {
            final Number number = (Number) value.raw();
            if (number instanceof Double) {
                return LzrNumber.of(number.doubleValue() + 1);
            }
            if (number instanceof Float) {
                return LzrNumber.of(number.floatValue() + 1);
            }
            if (number instanceof Long) {
                return LzrNumber.of(number.longValue() + 1);
            }
        }
        return LzrNumber.of(value.asInt() + 1);
    }
    
    private LzrValue decrement(LzrValue value) {
        if (value.type() == Types.NUMBER) {
            final Number number = (Number) value.raw();
            if (number instanceof Double) {
                return LzrNumber.of(number.doubleValue() - 1);
            }
            if (number instanceof Float) {
                return LzrNumber.of(number.floatValue() - 1);
            }
            if (number instanceof Long) {
                return LzrNumber.of(number.longValue() - 1);
            }
        }
        return LzrNumber.of(value.asInt() - 1);
    }
    
    private LzrValue negate(LzrValue value) {
        if (value.type() == Types.STRING) {
            final StringBuilder sb = new StringBuilder(value.asString());
            return new LzrString(sb.reverse().toString());
        }
        if (value.type() == Types.NUMBER) {
            final Number number = (Number) value.raw();
            if (number instanceof Double) {
                return LzrNumber.of(-number.doubleValue());
            }
            if (number instanceof Float) {
                return LzrNumber.of(-number.floatValue());
            }
            if (number instanceof Long) {
                return LzrNumber.of(-number.longValue());
            }
        }
        return LzrNumber.of(-value.asInt());
    }
    
    private LzrValue complement(LzrValue value) {
        if (value.type() == Types.NUMBER) {
            final Number number = (Number) value.raw();
            if (number instanceof Long) {
                return LzrNumber.of(~number.longValue());
            }
        }
        return LzrNumber.of(~value.asInt());
    }
    
    private LzrValue not(LzrValue value) {
        return LzrNumber.fromBoolean(value.asInt() == 0);
    }
    
    @Override
    public void accept(Visitor visitor) {
        visitor.visit(this);
    }
    
    @Override
    public <R, T> R accept(ResultVisitor<R, T> visitor, T t) {
        return visitor.visit(this, t);
    }
    
    @Override
    public String toString() {
        switch (operation) {
            case INCREMENT_POSTFIX:
            case DECREMENT_POSTFIX:
                return String.format("%s %s", expr1, operation);
            default:
                return String.format("%s %s", operation, expr1);
        }
    }
}

File: src/main/java/com/kingmang/lazurite/parser/ast/expressions/UnaryExpression.kt
package com.kingmang.lazurite.parser.ast.expressions

import com.kingmang.lazurite.core.Types
import com.kingmang.lazurite.parser.ast.Accessible
import com.kingmang.lazurite.parser.ast.statements.Statement
import com.kingmang.lazurite.patterns.visitor.ResultVisitor
import com.kingmang.lazurite.patterns.visitor.Visitor
import com.kingmang.lazurite.runtime.values.LzrNumber
import com.kingmang.lazurite.runtime.values.LzrNumber.Companion.fromBoolean
import com.kingmang.lazurite.runtime.values.LzrNumber.Companion.of
import com.kingmang.lazurite.runtime.values.LzrString
import com.kingmang.lazurite.runtime.values.LzrValue

data class UnaryExpression(
    val operation: Operator,
    val expr1: Expression
) : Expression, Statement {
    override fun execute() {
        eval()
    }

    override fun eval(): LzrValue {
        val value = this.expr1.eval()
        return when (this.operation) {
            Operator.INCREMENT_PREFIX -> {
                if (this.expr1 is Accessible)
                    this.expr1.set(increment(value))
                else increment(value)
            }

            Operator.DECREMENT_PREFIX -> {
                if (this.expr1 is Accessible)
                    this.expr1.set(decrement(value))
                else decrement(value)
            }

            Operator.INCREMENT_POSTFIX -> {
                if (this.expr1 is Accessible) {
                    this.expr1.set(increment(value))
                    value
                } else increment(value)
            }

            Operator.DECREMENT_POSTFIX -> {
                if (this.expr1 is Accessible) {
                    this.expr1.set(decrement(value))
                    value
                } else decrement(value)
            }

            Operator.NEGATE -> negate(value)
            Operator.COMPLEMENT -> complement(value)
            Operator.NOT -> not(value)
        }
    }

    private fun increment(value: LzrValue): LzrValue {
        if (value.type() == Types.NUMBER) {
            when (val number = (value as LzrNumber).raw()) {
                is Double -> return of(number.toDouble() + 1)
                is Float -> return of(number.toFloat() + 1)
                is Long -> return of(number.toLong() + 1)
            }
        }
        return of(value.asInt() + 1)
    }

    private fun decrement(value: LzrValue): LzrValue {
        if (value.type() == Types.NUMBER) {
            when (val number = (value as LzrNumber).raw()) {
                is Double -> return of(number.toDouble() - 1)
                is Float -> return of(number.toFloat() - 1)
                is Long -> return of(number.toLong() - 1)
            }
        }
        return of(value.asInt() - 1)
    }

    private fun negate(value: LzrValue): LzrValue =
        when (value.type()) {
            Types.STRING -> {
                val sb = StringBuilder(value.asString())
                LzrString(sb.reverse().toString())
            }

            Types.NUMBER -> {
                when (val number = (value as LzrNumber).raw()) {
                    is Double -> of(-number.toDouble())
                    is Float -> of(-number.toFloat())
                    is Long -> of(-number.toLong())
                    else -> of(-value.asInt())
                }
            }

            else -> of(-value.asInt())
        }

    private fun complement(value: LzrValue): LzrValue {
        if (value.type() == Types.NUMBER) {
            val number = (value as LzrNumber).raw()
            if (number is Long) {
                return of(number.toLong().inv())
            }
        }
        return of(value.asInt().inv())
    }

    private fun not(value: LzrValue): LzrValue =
        fromBoolean(value.asInt() == 0)

    override fun accept(visitor: Visitor) =
        visitor.visit(this)

    override fun <R, T> accept(visitor: ResultVisitor<R, T>, input: T): R? =
        visitor.visit(this, input)

    override fun toString(): String =
        when (this.operation) {
            Operator.INCREMENT_POSTFIX, Operator.DECREMENT_POSTFIX -> String.format("%s %s", expr1, this.operation)
            else -> String.format("%s %s", this.operation, expr1)
        }


    enum class Operator(val text: String) {
        INCREMENT_PREFIX("++"),
        DECREMENT_PREFIX("--"),
        INCREMENT_POSTFIX("++"),
        DECREMENT_POSTFIX("--"),
        NEGATE("-"),

        // Boolean
        NOT("!"),

        // Bitwise
        COMPLEMENT("~");

        override fun toString(): String =
            this.text
    }
}
